src/docfx/lib/git/GitUtility.cs

```diff
@@ -47,49 +47,76 @@ public static bool IsGitRepository(string path)
 
             if (git_repository_open(out var pRepo, repoPath) != 0)
             {
                 throw new ArgumentException($"Invalid git repo {repoPath}");
             }
 
-            var remotes = default(git_strarray);
-            if (git_remote_list(&remotes, pRepo) == 0)
-            {
-                var pstr = remotes.strings;
-                for (var i = 0; i < remotes.count; i++)
-                {
-                    var pName = *pstr++;
-                    var name = Marshal.PtrToStringUTF8(pName);
-                    if (git_remote_lookup(out var pRemote, pRepo, pName) == 0)
-                    {
-                        remote = Marshal.PtrToStringUTF8(git_remote_url(pRemote));
-                        git_remote_free(pRemote);
-                        if (name == "origin")
-                        {
-                            break;
-                        }
-                    }
-                }
-                git_strarray_free(&remotes);
-            }
-
             if (git_repository_head(out var pHead, pRepo) == 0)
             {
                 commit = git_reference_target(pHead)->ToString();
                 if (git_branch_name(out var pName, pHead) == 0)
                 {
                     branch = Marshal.PtrToStringUTF8(pName);
                 }
+
+                remoteName = GetUpstreamRemoteName(pRepo, pHead);
                 git_reference_free(pHead);
             }
 
+            url = GetRepositoryUrl(pRepo, remoteName ?? "origin");
+
             git_repository_free(pRepo);
 
-            return (remote, branch, commit);
+            return (url, branch, commit);
+
+            static unsafe string? GetUpstreamRemoteName(IntPtr pRepo, IntPtr pBranch)
+            {
+                string? result = null;
+                if (git_branch_upstream(out var pUpstream, pBranch) == 0)
+                {
+                    git_buf buf;
+                    var pUpstreamName = git_reference_name(pUpstream);
+                    if (git_branch_remote_name(&buf, pRepo, pUpstreamName) == 0)
+                    {
+                        result = Marshal.PtrToStringUTF8(buf.ptr) ?? "origin";
+                        git_buf_free(&buf);
+                    }
+                    git_reference_free(pUpstream);
+                }
+                return result;
+            }
+
+            static unsafe string? GetRepositoryUrl(IntPtr pRepo, string remoteName)
+            {
+                var result = (string?)null;
+                var remotes = default(git_strarray);
+                if (git_remote_list(&remotes, pRepo) == 0)
+                {
+                    var pstr = remotes.strings;
+                    for (var i = 0; i < remotes.count; i++)
+                    {
+                        var pName = *pstr++;
+                        var name = Marshal.PtrToStringUTF8(pName);
+                        if (git_remote_lookup(out var pRemote, pRepo, pName) == 0)
```

Comment:
>git_remote_lookup(out var pRemote, pRepo, pName) == 0 [](start = 28, length = 53)

if ((result == null || name == remoteName) && git_remote_lookup(out var pRemote, pRepo, pName) == 0)
// only need to store the first found remote, and then search for `remoteName`

Reply:
I was lazy